protocol/x/vault/keeper/withdraw.go

```diff
@@ -117,3 +115,164 @@ func (k Keeper) GetVaultWithdrawalSlippage(
 		estimatedSlippage,
 	), nil
 }
+
+// WithdrawFromMegavault withdraws from megavault to a subaccount.
+func (k Keeper) WithdrawFromMegavault(
+	ctx sdk.Context,
+	toSubaccount satypes.SubaccountId,
+	sharesToWithdraw *big.Int,
+	minQuoteQuantums *big.Int,
+) (redeemedQuoteQuantums *big.Int, err error) {
+	// 1. Check that the owner is withdrawing less that or equal to their unlocked shares.
+	ownerShares, exists := k.GetOwnerShares(ctx, toSubaccount.Owner)
+	if !exists {
+		return nil, types.ErrOwnerNotFound
+	}
+	ownerShareUnlocks, _ := k.GetOwnerShareUnlocks(ctx, toSubaccount.Owner)
+	ownerSharesAfterWithdrawal := ownerShares.NumShares.BigInt()
+	ownerSharesAfterWithdrawal.Sub(ownerSharesAfterWithdrawal, sharesToWithdraw)
+	ownerLockedShares := ownerShareUnlocks.GetTotalLockedShares()
+	if ownerSharesAfterWithdrawal.Cmp(ownerLockedShares) < 0 {
+		return nil, errorsmod.Wrapf(
+			types.ErrInsufficientWithdrawableShares,
+			"shares to withdraw: %s, owner total shares: %s, owner locked shares: %s",
+			sharesToWithdraw,
+			ownerShares,
+			ownerLockedShares,
+		)
+	}
+
+	// 2. Redeem from main vault.
+	totalShares := k.GetTotalShares(ctx).NumShares.BigInt()
+	redeemedQuoteQuantums, err = k.GetSubaccountEquity(ctx, types.MegavaultMainSubaccount)
+	if err != nil {
+		log.ErrorLogWithError(ctx, "Failed to get megavault main vault equity", err)
+		return nil, err
+	}
+	redeemedQuoteQuantums.Mul(redeemedQuoteQuantums, sharesToWithdraw)
+	redeemedQuoteQuantums.Quo(redeemedQuoteQuantums, totalShares)
+
+	// 3. Redeem from each sub vault.
+	vaultParamsIterator := k.getVaultParamsIterator(ctx)
+	defer vaultParamsIterator.Close()
+	for ; vaultParamsIterator.Valid(); vaultParamsIterator.Next() {
+		vaultId, err := types.GetVaultIdFromStateKey(vaultParamsIterator.Key())
+		if err != nil {
+			log.ErrorLogWithError(ctx, "Failed to get vault ID from state key", err)
+			continue
+		}
+		var vaultParams types.VaultParams
+		k.cdc.MustUnmarshal(vaultParamsIterator.Value(), &vaultParams)
+
+		_, perpetual, marketParam, marketPrice, err := k.GetVaultClobPerpAndMarket(ctx, *vaultId)
+		if err != nil {
+			log.ErrorLogWithError(ctx, "Failed to get perpetual and market", err, "Vault ID", vaultId)
```

Comment:
nit: These errors maybe should also surface that they happening during withdrawal?